### PR TITLE
Adds InputSocket & OutputSocket

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Since input sockets and output sockets have different functionality and
need to be distinguished they have been seperated into different
classes.

### CODE REVIEW

**Comment by ThomasJClark** [comment]
Also, input and output sockets should be different controls

**Comment by JLLeitschuh** [comment]
Please move your comments into the [file changes](https://github.com/WPIRoboticsProjects/GRIP/pull/29/files) so I can comment on them and they will show up in this PR.

**Comment by ThomasJClark** [comment]
Which comments?  I all of my comments on individual lines are showing up here.

**Comment by JLLeitschuh** [comment]
I don't see them.

**Comment by JLLeitschuh** [comment]
I don't understand what you mean by different controls.

File: src/main/java/edu/wpi/grip/ui/SocketHandleView.java [hunk 1]
```diff
@@ -88,29 +85,20 @@ public SocketHandleView(EventBus eventBus, SocketControlView controlView) {
...
-                // Check which socket was the input and which was the output.  The user can create a connection by
-                // dragging either into the other.
-                if (other.isInputSocket()) {
-                    inputSocket = other.getSocket();
-                    outputSocket = this.controlView.getSocket();
-                } else {
-                    inputSocket = this.controlView.getSocket();
-                    outputSocket = other.getSocket();
+                try {
+                    Connection connection = this.controlView.getSocket().connect(other.getSocket());
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: This hides the implementation details from the UI. The UI doesn't need to know how a socket connects just that it does connect.

File: src/main/java/edu/wpi/grip/ui/SocketHandleView.java [hunk 2]
```diff
@@ -88,29 +85,20 @@ public SocketHandleView(EventBus eventBus, SocketControlView controlView) {
...
-                @SuppressWarnings("unchecked")
-                Connection connection = new Connection(this.eventBus, outputSocket, inputSocket);
-                eventBus.post(new ConnectionAddedEvent(connection));
             });
         });
 
         // Accept input sockets being dragged onto output sockets and vice versa.
         this.setOnDragOver(dragEvent -> {
             draggingSocket.ifPresent(other -> {
-                if (other.isInputSocket() != this.controlView.isInputSocket()) {
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: This exposes implementation details.

File: src/main/java/edu/wpi/grip/ui/SocketHandleView.java [hunk 3]
```diff
@@ -88,29 +85,20 @@ public SocketHandleView(EventBus eventBus, SocketControlView controlView) {
...
+                if (other.canAcceptTransfer(this.controlView)) {
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: This hides it and allows this rule to change.

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: No, sockets know nothing about connections (or "transfers" as you've apparently started calling them).  We've already defined the role of the socket class, let's not tack on more functionality that really shouldn't be in there.

File: src/main/java/edu/wpi/grip/ui/SocketHandleView.java (hunk 2 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: No, this is a UI thing.  Since you get an onDragOver as soon as you start dragging a socket, we need to make sure that the user has dragged it away to another socket before we actually handle those events.

File: src/main/java/edu/wpi/grip/ui/SocketHandleView.java (hunk 1 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/SocketHandleView.java]: Sockets don't need to know how they connect, `Connection` is supposed to do that.

**Comment by ThomasJClark** [comment]
> Also, I don't understand what you mean by different controls.
I mean that there should be one JavaFX control that represents input sockets and one for output sockets, instead of one `SocketControlView` that takes either and changes between two different controls.

File: src/main/java/edu/wpi/grip/core/OutputSocket.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+     * @param eventBus   The Guava {@link EventBus} used by the application.
+     * @param socketHint {@link #getSocketHint}
+     */
+    public OutputSocket(EventBus eventBus, SocketHint<T> socketHint){
+        super(eventBus, socketHint);
+    }
+
+    @Override
+    public boolean setValue(T value){
+        boolean valueSet = super.setValue(value);
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/OutputSocket.java]: Move this. Remove boolean.

File: src/main/java/edu/wpi/grip/core/Step.java [hunk 5]
```diff
@@ -53,21 +46,21 @@ public Operation getOperation() {
...
-    public Socket<?>[] getOutputSockets() {
+    public OutputSocket<?>[] getOutputSockets() {
         return outputSockets;
     }
 
     @Subscribe
     public void onInputSocketChanged(SocketChangedEvent e) {
         // If this socket that changed is one of the inputs to this step, run the operation with the new value.
-        if (e.getSocket().getStep() == this) {
+        if (Arrays.asList(inputSockets).contains(e.getSocket())) {
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/Step.java]: Revert this change.  This has absolutely nothing to do with the rest of the commit.

File: src/main/java/edu/wpi/grip/ui/InputSocketView.java [hunk 6]
```diff
@@ -25,7 +26,7 @@
 import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
- * A JavaFX control that renders a {@link Socket} that is an input to a step.  This includes an identifier, a handle
+ * A JavaFX control that renders a {@link InputSocket} that is an input to a step.  This includes an identifier, a handle
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/InputSocketView.java]: a -> an

File: src/main/java/edu/wpi/grip/core/Step.java (hunk 5 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/Step.java]: You want it in a separate PR or separate commit in this PR?

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/Step.java]: Nope

File: src/main/java/edu/wpi/grip/core/Socket.java [hunk 7]
```diff
@@ -13,38 +12,39 @@
...
+ * Sockets that are given to operations are {@link InputSocket InputSockets}, and sockets that operations store their
+ * results are {@link OutputSocket OutputSockets}.
+ *
+ * @param <T> The type of the value that this socket stores
  */
-public class Socket<T> {
+public abstract class Socket<T> {
     public enum Direction {INPUT, OUTPUT}
 
     private final EventBus eventBus;
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/Socket.java]: This should be made protected, right?  For when subclasses need to send events. (This isn't leaking anything since almost every class in the application has access to the same eventbus)

Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/Socket.java]: Eh, I thought about that but this is better practice. If I make it protected then other objects in the package can access the variable.

File: src/main/java/edu/wpi/grip/core/Step.java (hunk 5 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/Step.java]: Nope to which?

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/Step.java]: Both.  I mean, I guess you can do a PR if you want to, but that doesn't mean it'll get merged.

File: src/main/java/edu/wpi/grip/core/Socket.java [hunk 8]
```diff
@@ -207,9 +152,7 @@ public String toString() {
         return MoreObjects.toStringHelper(this)
                 .add("socketHint", getSocketHint())
                 .add("value", getValue())
-                .add("published", isPublished())
-                .add("previewed", isPreviewed())
-                .add("connections", getConnections())
+                .add("direction", getDirection())
                 .toString();
     }
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/Socket.java]: InputSocket doesn't overload this method.

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/Socket.java]: ok

File: src/main/java/edu/wpi/grip/core/OutputSocket.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        super(eventBus, socketHint, Direction.OUTPUT);
+        this.eventBus = eventBus;
+    }
+
+    @Override
+    public void setValue(T value){
+        // Check equality first before the value is ste
+        final boolean isNewValue = !this.getValue().equals(value);
+        // Set the value now that we know if there is a new value being stored
+        super.setValue(value);
```
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/OutputSocket.java]: It is impossible to set the value of the parent class without calling super.

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/OutputSocket.java]: `value` and `eventBus` should be proected because the subclasses need to access them.  Your workaround is just a hack to intercept the private variables, it's not better design because you avoid the `protected` keyword.

I would definitely just make the fields  `protected` and avoid all of this.

Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/OutputSocket.java]: Protected means package access. Not what you want.
